packages/@lwc/module-resolver/src/index.ts

```diff
@@ -4,117 +4,147 @@
+            return createRegistryEntry(entry, specifier, opts);
+        }
+    }
 }
 
-function resolveModulesFromDir(modulesDir: string): RegistryEntry[] {
-    const namespaces = fs.readdirSync(modulesDir);
-    const resolvedModules: RegistryEntry[] = [];
-    namespaces.forEach(ns => {
-        if (ns[0] !== '.' && fs.lstatSync(path.join(modulesDir, ns)).isDirectory()) {
-            const namespacedModuleDir = path.join(modulesDir, ns);
-            const modules = fs.readdirSync(namespacedModuleDir);
-            modules.forEach(moduleName => {
-                const moduleDir = path.join(namespacedModuleDir, moduleName);
-                const entry = getModuleEntry(moduleDir, moduleName);
-                if (entry) {
-                    const specifier = `${ns}/${moduleName}`;
-                    resolvedModules.push({ entry, specifier });
+function resolveModuleFromNpm(
+    specifier: string,
+    npmModuleRecord: NpmModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { npm, map: aliasMapping } = npmModuleRecord;
+    const pkgJsonPath = require.resolve(`${npm}/package.json`, { paths: [opts.rootDir] });
+    const packageDir = path.dirname(pkgJsonPath);
+    const lwcConfig = getLwcConfig(packageDir);
+    const modules = lwcConfig && lwcConfig.modules;
+
+    if (modules) {
+        validateNpmConfig(lwcConfig);
+        const exposedModules = lwcConfig.expose || [];
+        if (exposedModules.includes(specifier)) {
+            for (const moduleRecord of modules) {
+                if (!isNpmModuleRecord(moduleRecord)) {
+                    const registryEntry = resolveModuleRecordType(specifier, moduleRecord, {
+                        rootDir: packageDir,
+                    });
+                    if (registryEntry) {
+                        if (aliasMapping && aliasMapping[specifier]) {
+                            registryEntry.specifier = aliasMapping[specifier];
+                        }
+                        return registryEntry;
+                    }
                 }
-            });
+            }
+
+            throw new Error(`Unable to find ${specifier} under package ${npmModuleRecord.npm}`);
         }
-    });
+    }
+}
 
-    return resolvedModules;
+function resolveModuleRecordType(
+    specifier: string,
+    moduleRecord: ModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { rootDir } = opts;
+    validateModuleRecord(moduleRecord);
+
+    if (isAliasModuleRecord(moduleRecord) && moduleRecord.name === specifier) {
+        return resolveModuleFromAlias(moduleRecord, { rootDir });
+    } else if (isDirModuleRecord(moduleRecord)) {
+        return resolveModuleFromDir(specifier, moduleRecord, { rootDir });
+    } else if (isNpmModuleRecord(moduleRecord)) {
+        return resolveModuleFromNpm(specifier, moduleRecord, opts);
+    } else {
+        throw new Error(`Invalid moduleRecord type ${moduleRecord}`);
```

Comment:
This case should be covered by `validateModuleRecord`.

Reply:
I think having this `throw` actually causes a bug and needs to be removed.

What if I modify the test the module resolution definition (https://github.com/salesforce/lwc/pull/1602/files#diff-1517212a5967f0fe035a2b279c690b1bR1) to include some other record:

```js
{
    "modules": [
        {
          "name": "some/button", // these are made up
          "path": "src/modules/root/some/button.js"
        },
        { "dir" : "src/modules" },
        { "npm" :"fancy-components" },
        { "npm" : "@ui/components" }
    ]
}
```

The `some/button` record is made up, it isn't used anywhere. The point is though, when `fancy/bar` is resolved, the first record is defined, and passes validation within `validateModuleRecord`. But because `moduleRecord.name !== specifier` the code proceeds to line 116 and throws an error.

Reply:
@pmdartus I'm happy to add a more exhaustive tests that cover a variety of scenarios, including the above use case.